aws.go

```
@@ -1,96 +1,210 @@
 			}
-			fileConf = iniFile[n]
-			if fileConf != nil {
-				break
+		}
+		creds = loadCachedCreds(profile)
+		if creds == nil || creds.IsExpired() {
+			if profile.MfaSerial != "" {
+				profile.Token = readToken()
+				creds = getStsCredentials(profile)
+			} else {
+				creds = credentials.NewStaticCredentials(profile.AwsAccessKeyId, profile.AwsSecretAccessKey, "")
 			}
 		}
+		sess = session.New(&aws.Config{Credentials: creds, Region: &profile.Region})
+	}
 
-		if fileConf == nil {
-			err = errors.New(
-				fmt.Sprintf("Couldn't find a suitable AWS config profile; looked for profiles named '%s'. Please add one to your AWS config file at %s",
-					strings.Join(profiles, "', '"), confFn))
-			return
-		}
+	return
+}
+
+func getStsCredentials(profile Profile) (creds *credentials.Credentials) {
+	staticCreds := credentials.NewStaticCredentials(profile.AwsAccessKeyId, profile.AwsSecretAccessKey, "")
+	staticCreds.Get()
+	client := sts.New(session.New(&aws.Config{Credentials: staticCreds, Region: &profile.Region}))
+
+	sessionName := "AWS-Profile-session-" + strconv.Itoa(int(time.Now().Unix()))
+	input := sts.AssumeRoleInput{
+		RoleArn:         &profile.RoleArn,
+		SerialNumber:    &profile.MfaSerial,
+		RoleSessionName: &sessionName,
+		TokenCode:       &profile.Token,
+	}
 
-		conf.AccessKey = fileConf["aws_access_key_id"]
-		conf.SecretKey = fileConf["aws_secret_access_key"]
-		conf.Region = aws.Regions[fileConf["region"]]
+	output, err := client.AssumeRole(&input)
+	if err != nil {
+		log.Fatal(err)
 	}
+	saveCachedCreds(profile, output)
 
-	envRegion := os.Getenv("AWS_DEFAULT_REGION")
-	if envRegion != "" {
-		conf.Region = aws.Regions[envRegion]
+	creds = credentials.NewStaticCredentials(*output.Credentials.AccessKeyId, *output.Credentials.SecretAccessKey, *output.Credentials.SessionToken)
+	creds.Get()
+	return
+}
+
+func readToken() (token string) {
+	var err error
+	for {
+		fmt.Print("Enter MFA code: ")
+		fmt.Scanln(&token)
```

This ignores the error possibly returned